### PR TITLE
Test framework for the PAB effects

### PR DESCRIPTION
**Pull request by gege251**
This PR includes a framework for testing the PAB. I changed most IO actions to freer effects, so we can run them as a pure function in the tests, and inspect the cardano-cli calls, and file read/write actions.

I will still need to write more tests and maybe refactor the framework to reduce boilerplate.

Solves: https://github.com/mlabs-haskell/mlabs-pab/issues/3

### CODE REVIEW

File: test/Spec/MLabsPAB/Contract.hs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+-}
+tests :: TestTree
+tests =
+  testGroup
+    "MLabsPAB.Contracts"
+    [ testCase "Send ada to address" sendAda
+    , testCase "Support multiple signatories" multisigSupport
+    ]
+
+sendAda :: Assertion
```
Comment (samuelWilliams99) [inline test/Spec/MLabsPAB/Contract.hs:34]: Think we could do some property testing in this module?

Comment (samuelWilliams99) [reply test/Spec/MLabsPAB/Contract.hs:34]: Also can we test for non Ada values

**Review by samuelWilliams99 — COMMENTED** [review]
I see theres a few universal quantifiers without explicit `forall`s, we should get those in, and add Kind signatures to them.

Otherwise - Very impressive! Freer looks super cool

**Comment by gege251** [comment]
Unfortunately I couldn't add property based tests yet, it would be a little bit more involved, and I still don't know how much this project would be used after the real pab is ready...

File: src/MLabsPAB/PreBalance.hs [hunk 2]
```diff
@@ -158,7 +159,9 @@ balanceNonAdaOuts ownPkh utxos tx =
             txOuts
           (txOut@TxOut {txOutValue = v} : txOuts, txOuts') ->
             txOut {txOutValue = v <> nonAdaChange} : (txOuts <> txOuts')
-   in Right $ tx {txOutputs = outputs}
+   in if Value.isZero nonAdaChange
```
Comment (samuelWilliams99) [inline src/MLabsPAB/PreBalance.hs:162]: It doesn't look like this function can fail, does it need to be in an Either context?

File: test/Spec/MLabsPAB/Contract.hs [hunk 3]
```diff
@@ -29,108 +49,278 @@ tests =
...
+       --mint-redeemer-file result-scripts/redeemer-${redeemerHash}.json
+       --mint 5 ${curSymbol'}.testToken
+       --change-address ${addr1}
+       --required-signer signing-keys/signing-key-${pkh1'}.skey
+       --mainnet --protocol-params-file ./protocol.json --out-file tx.raw
+      |]
+
+  assertFiles
+    state
+    [ [text|result-scripts/policy-${curSymbol'}.plutus|]
```
Comment (samuelWilliams99) [inline test/Spec/MLabsPAB/Contract.hs:233]: Can we assert the content of these files or is it not worth it?

File: test/Spec/MockContract.hs [hunk 4]
```diff
@@ -159,12 +198,75 @@ runPABEffectPure config initState req =
...
+
+  let network = state.contractEnv.cePABConfig.pcNetwork
+  pure $
+    mockQueryUtxoOut $
+      filter
+        ((==) addr . unsafeSerialiseAddress network . Ledger.txOutAddress . snd)
+        state.utxos
+
+mockQueryUtxoOut :: [(TxOutRef, TxOut)] -> String
+mockQueryUtxoOut utxos =
```
Comment (samuelWilliams99) [inline test/Spec/MockContract.hs:232]: This function has *quite* a lot of scope, could we split it up to be a little less massive?

File: test/Spec/MockContract.hs [hunk 5]
```diff
@@ -176,33 +278,44 @@ mockThreadDelay :: Int -> MockContract ()
...
+        mapLeft (FileError filepath) $ deserialiseFromTextEnvelope ttoken te
+      Just _ -> Left $ FileError filepath $ TextEnvelopeAesonDecodeError "Invalid format."
 
-  pure $ do
-    te <-
-      maybeToRight
-        (FileIOError filepath (IOError Nothing NoSuchThing "" "No such file in the MockContractState" Nothing Nothing))
-        maybeFile
-    mapLeft (FileError filepath) $ deserialiseFromTextEnvelope ttoken te
+-- pure $ do
```
Comment (samuelWilliams99) [inline test/Spec/MockContract.hs:296]: If not needed, lets remove

**Review by samuelWilliams99 — CHANGES_REQUESTED** [review]

File: src/MLabsPAB/PreBalance.hs (hunk 2 above)
Comment (gege251) [reply src/MLabsPAB/PreBalance.hs:162]: Well, I just didn't want to break the pattern here:
https://github.com/mlabs-haskell/mlabs-pab/blob/e99af5435701a87f1df8791cd96e94d0e90355f7/src/MLabsPAB/PreBalance.hs#L50-L54
But yeah, you're right, I'm going to remove this

File: test/Spec/MLabsPAB/Contract.hs (hunk 3 above)
Comment (gege251) [reply test/Spec/MLabsPAB/Contract.hs:233]: Interesting, I have a feeling that would take more time than it's worth, but definitely a good idea.
Maybe in a later iteration (if there is one)

File: test/Spec/MockContract.hs (hunk 4 above)
Comment (gege251) [reply test/Spec/MockContract.hs:232]: Absolutely

File: test/Spec/MockContract.hs [hunk 6]
```diff
@@ -268,6 +249,24 @@ mockQueryUtxoOut utxos =
...
+            if curSymbol == Ada.adaSymbol
+              then "lovelace"
+              else
+                let curSymbol' =
+                      encodeByteString $
+                        fromBuiltin $ Value.unCurrencySymbol curSymbol
+                    tokenName' =
+                      decodeUtf8 $
+                        fromBuiltin $ Value.unTokenName tokenName
+                 in [text|${curSymbol'}.${tokenName'}|]
```
Comment (samuelWilliams99) [inline test/Spec/MockContract.hs:267]: If the token name is empty, the dot shouldn't be there

Comment (gege251) [reply test/Spec/MockContract.hs:267]: Aaa, true, I absolutely forgot! Thanks

Comment (gege251) [reply test/Spec/MockContract.hs:267]: I'll write tests for this

Comment (samuelWilliams99) [reply test/Spec/MockContract.hs:267]: Nice, thanks :)

Comment (gege251) [reply test/Spec/MockContract.hs]: I realised that I haven't backported your changes from the bulk airdrop repo. I fixed that. Also, I had to fix the UtxoParser too, so I ended up writing some tests for the parser.

**Review by samuelWilliams99 — CHANGES_REQUESTED** [review]
Could this fail for addresses of validators, if the tokens datum isn't None?

File: src/MLabsPAB/PreBalance.hs [hunk 7]
```diff
@@ -180,13 +181,13 @@ addSignatories ownPkh privKeys pkhs tx =
 showText :: forall (a :: Type). Show a => a -> Text
 showText = Text.pack . show
 
+-- | Filter by key for Associated maps (why doesn't this exist?)
+filterKey :: (k -> Bool) -> AssocMap.Map k v -> AssocMap.Map k v
+filterKey f = AssocMap.mapMaybeWithKey $ \k v -> if f k then Just v else Nothing
```
[the remark this answers is not in the document]
Comment (gege251) [reply src/MLabsPAB/PreBalance.hs:186]: yeah, sorry, I plagiarized your work

File: src/MLabsPAB/CardanoCLI.hs [hunk 8]
```diff
@@ -91,11 +91,12 @@ utxosAt pabConf address = do
             , networkOpt pabConf
             ]
       , cmdOutParser =
-          Map.fromList . mapMaybe (rightToMaybe . toUtxo) . drop 2 . Text.lines . Text.pack
+          Map.fromList
+            . fromMaybe []
+            . rightToMaybe
```
[the remark this answers is not in the document]
Comment (gege251) [reply src/MLabsPAB/CardanoCLI.hs:96]: yeah, that was silly

File: src/MLabsPAB/UtxoParser.hs [hunk 9]
```diff
@@ -48,14 +50,15 @@ utxoMapParser address = do
 
 skipLine :: Int -> Parser ()
 skipLine n =
-  forM_ [1 .. n] $ \_ -> do
-    skipWhile (not . inClass "\r\n")
-    skipWhile $ inClass "\r\n"
+  void $
+    count n $ do
+      skipWhile (not . isEndOfLine)
+      skipWhile $ inClass "\r\n"
```
Comment (samuelWilliams99) [inline src/MLabsPAB/UtxoParser.hs:56]: `isEndOfLine` here as well?

**Review by samuelWilliams99 — APPROVED** [review]

**Comment by gege251** [comment]
> Could this fail for addresses of validators, if the tokens datum isn't None?

I forgot to reply to this: when the queried address is a script address, we there must be datumhash for the utxos. on the other hand, utxos sitting at pub key addresses cannot have datums, so its a simple mapping. There is a test for script address utxos with datumhashes.
btw, it is possible to create a utxo that is sitting at a script address, but don't have any datum hashes, but it is impossible to redeem it, and there isn't a way to represent it with plutus types.

**Pull request merged**